### PR TITLE
Feature/defer stats

### PR DESCRIPTION
**Pull request by tuj**
Deferring stats logging til after response has been sent.

### CODE REVIEW

File: src/Api/DataProvider/AbstractElasticSearchDataProvider.php [hunk 1]
```diff
@@ -180,14 +180,21 @@ protected function logStatistics(string $type, array $identifiers, array $result
...
+            KernelEvents::TERMINATE,
+            function (TerminateEvent $event) use ($className, $type, $identifiers, $results, $request) {
+                $this->statsLogger->info('Cover request/response', [
+                    'service' => $className,
+                    // @TODO Log clientID when authentication implemented, log 'REST_API' for now to allow stats filtering on REST.
+                    'clientID' => 'REST_API',
+                    'remoteIP' => $request->getClientIp(),
+                    'isType' => $type,
+                    'isIdentifiers' => $identifiers,
+                    'fileNames' => $this->getImageUrls($results),
```
Comment (cableman) [inline src/Api/DataProvider/AbstractElasticSearchDataProvider.php]: Will this `this` reference work in an event outside the scope of the object will `this` not point to something else?

File: src/Service/MoreInfoService/AbstractMoreInfoService.php [hunk 2]
```diff
@@ -237,14 +237,21 @@ public function moreInfo($body): MoreInfoResponse
...
+        // delivered.
+        $this->dispatcher->addListener(
+            KernelEvents::TERMINATE,
+            function (TerminateEvent $event) use ($body, $searchParameters, $results) {
+                $this->statsLogger->info('Cover request/response', [
+                    'service' => 'MoreInfoService',
+                    'clientID' => $body->authentication->authenticationGroup,
+                    'remoteIP' => $this->requestStack->getCurrentRequest()->getClientIp(),
+                    'searchParameters' => $searchParameters,
+                    'fileNames' => $this->getImageUrls($results),
```
Comment (cableman) [inline src/Service/MoreInfoService/AbstractMoreInfoService.php]: Same about `this` here as above.

File: src/Service/StatsLoggingService.php [hunk 3]
```diff
@@ -0,0 +1,144 @@
+<?php
+
```
Comment (cableman) [inline src/Service/StatsLoggingService.php]: Missing @file description

File: src/Service/MoreInfoService/AbstractMoreInfoService.php [hunk 4]
```diff
@@ -83,8 +83,8 @@ abstract class AbstractMoreInfoService extends SoapClient
      *
      * @param Type $index
      *   Elastica index
-     * @param LoggerInterface $statsLogger
-     *   Statistics logger
+     * @param StatsLoggingService $statsLoggingService
```
Comment (cableman) [inline src/Service/MoreInfoService/AbstractMoreInfoService.php]: The logger service implements the `LoggerInterface` this should be the type that the function takes.

File: src/Controller/SoapController.php [hunk 5]
```diff
@@ -22,68 +23,83 @@ class SoapController extends AbstractController
...
+        return $this->soap($request, $defaultCoverMoreInfoService, $statsLoggingService, $dbcWsdlFile);
     }
 
     /**
      * Return a SOAP response for the given request.
      *
      * @param Request $request
      * @param AbstractMoreInfoService $moreInfoService
-     * @param LoggerInterface $statsLogger
+     * @param StatsLoggingService $statsLoggingService
```
Comment (cableman) [inline src/Controller/SoapController.php]: The logger service implements the `LoggerInterface` this should be the type that the function takes.

File: src/Controller/SoapController.php [hunk 6]
```diff
@@ -22,68 +23,83 @@ class SoapController extends AbstractController
...
+        return $this->soap($request, $dbcMoreInfoService, $statsLoggingService, $dbcWsdlFile);
     }
 
     /**
      * @Route("/defaultcover/2.11/", name="default_cover_soap")
      *
      * @param Request $request
      * @param DefaultCoverMoreInfoService $defaultCoverMoreInfoService
-     * @param LoggerInterface $statsLogger
+     * @param StatsLoggingService $statsLoggingService
```
Comment (cableman) [inline src/Controller/SoapController.php]: The logger service implements the `LoggerInterface` this should be the type that the function takes.

File: src/Controller/SoapController.php [hunk 7]
```diff
@@ -22,68 +23,83 @@ class SoapController extends AbstractController
...
+        return $this->soap($request, $ddbMoreInfoService, $statsLoggingService, $dbcWsdlFile);
     }
 
     /**
      * @Route("/fbs/2.11/", name="fbs_soap")
      *
      * @param Request $request
      * @param DbcMoreInfoService $dbcMoreInfoService
-     * @param LoggerInterface $statsLogger
+     * @param StatsLoggingService $statsLoggingService
```
Comment (cableman) [inline src/Controller/SoapController.php]: The logger service implements the `LoggerInterface` this should be the type that the function takes.

File: src/Controller/SoapController.php [hunk 8]
```diff
@@ -22,68 +23,83 @@ class SoapController extends AbstractController
...
      * @param DdbMoreInfoService $ddbMoreInfoService
-     * @param LoggerInterface $statsLogger
+     * @param StatsLoggingService $statsLoggingService
+     *    Statistics logging service
      * @param $projectDir
      *
      * @return Response
      */
-    public function ddbSoap(Request $request, DdbMoreInfoService $ddbMoreInfoService, LoggerInterface $statsLogger, $projectDir): Response
+    public function ddbSoap(Request $request, DdbMoreInfoService $ddbMoreInfoService, StatsLoggingService $statsLoggingService, $projectDir): Response
```
Comment (cableman) [inline src/Controller/SoapController.php]: The logger service implements the `LoggerInterface` this should be the type that the function takes.

File: src/Api/DataProvider/AbstractElasticSearchDataProvider.php [hunk 9]
```diff
@@ -42,18 +42,18 @@ abstract class AbstractElasticSearchDataProvider
...
-     *   Logger for statistics
+     * @param StatsLoggingService $statsLoggingService
+     *   Statistics logging service
      * @param EventDispatcherInterface $dispatcher
      *   Symfony Event Dispatcher
      * @param IdentifierFactory $factory
      *   Factory to create Identifier Data Transfer Objects (DTOs)
      */
-    public function __construct(Type $index, RequestStack $requestStack, LoggerInterface $statsLogger, EventDispatcherInterface $dispatcher, IdentifierFactory $factory)
+    public function __construct(Type $index, RequestStack $requestStack, StatsLoggingService $statsLoggingService, EventDispatcherInterface $dispatcher, IdentifierFactory $factory)
```
Comment (cableman) [inline src/Api/DataProvider/AbstractElasticSearchDataProvider.php]: The logger service implements the `LoggerInterface` this should be the type that the function takes.

File: src/Service/StatsLoggingService.php [hunk 10]
```diff
@@ -0,0 +1,144 @@
...
+use Symfony\Component\EventDispatcher\EventDispatcherInterface;
+use Symfony\Component\HttpKernel\Event\TerminateEvent;
+use Symfony\Component\HttpKernel\KernelEvents;
+
+/**
+ * Class StatsLoggingService.
+ */
+class StatsLoggingService implements LoggerInterface
+{
+    const EMERGENCY = 'emergency';
```
Comment (cableman) [inline src/Service/StatsLoggingService.php]: Is these not defined in the `LoggerInterface`? or can be extended from some where?

File: src/Service/StatsLoggingService.php (hunk 3 above)
Comment (tuj) [reply src/Service/StatsLoggingService.php]: Fixed

File: src/Service/StatsLoggingService.php (hunk 10 above)
Comment (tuj) [reply src/Service/StatsLoggingService.php]: Fixed

File: src/Service/MoreInfoService/AbstractMoreInfoService.php (hunk 4 above)
Comment (tuj) [reply src/Service/MoreInfoService/AbstractMoreInfoService.php]: But this is autowired. Therefore it needs to be the precise service.

File: src/Controller/SoapController.php (hunk 5 above)
Comment (tuj) [reply src/Controller/SoapController.php]: But this is autowired. Therefore it needs to be the precise service.

File: src/Controller/SoapController.php (hunk 6 above)
Comment (tuj) [reply src/Controller/SoapController.php]: But this is autowired. Therefore it needs to be the precise service.

File: src/Controller/SoapController.php (hunk 7 above)
Comment (tuj) [reply src/Controller/SoapController.php]: But this is autowired. Therefore it needs to be the precise service.

File: src/Controller/SoapController.php (hunk 8 above)
Comment (tuj) [reply src/Controller/SoapController.php]: But this is autowired. Therefore it needs to be the precise service.

File: src/Api/DataProvider/AbstractElasticSearchDataProvider.php (hunk 9 above)
Comment (tuj) [reply src/Api/DataProvider/AbstractElasticSearchDataProvider.php]: But this is autowired. Therefore it needs to be the precise service.

File: src/Controller/SoapController.php (hunk 8 above)
Comment (tuj) [reply src/Controller/SoapController.php]: I have fixed it for the private function at the bottom of the SoapController.php

**Pull request merged**